### DISCUSSION TITLE
Allow internal struct enumerators and GetEnumerator to be used by foreach loop.

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [dotnet/csharplang#952] [General]
I don't see why following gives error. apparently foreach can not see the method.

```
        internal Enumerator GetEnumerator()
        {
            return new Enumerator(this);
        }
```

foreach will only see this method if its public. but why is this requirement? my type is public, but I want the struct enumerator to be internal. I know workaround for this, but I'm just being curios.

I don't think its CLR thing again, it simply can be understood by compiler, please allow this.

**Comment by HaloFour** [comment]
Changing this behavior might break the meaning of existing code unless the compiler favors implementations of `IEnumerable<T>` and `IEnumerable` over an `internal` enumerator method.

**Comment by yaakov-h** [comment]
@HaloFour with explicit interface implementations?

**Comment by HaloFour** [comment]
@yaakov-h 

> with explicit interface implementations?

Yes, although the CLR is not limited to how C# does explicit implementations.  For example, in VB.NET you're allowed to both rename the implemented method as well as change the accessibility to whatever you want:

```vb
Public Class Foo
    Implements IEnumerable(Of Integer)
    
    Private Function Bar() As IEnumerator(Of Integer) _
        Implements IEnumerable(Of Integer).GetEnumerator
        
        Return Nothing
    End Function
    
    Friend Function Baz() As IEnumerator _
        Implements IEnumerable.GetEnumerator
        
        Return Nothing
    End Function
End Class
```

In either case the C# compiler currently favors `IEnumerable` implementations if the compiler can't find a `public` method that matches the enumerable convention.  This is probably a minor point but it could lead to code being reinterpreted at recompile if an `internal` method (or any other accessibility modifier) is considered as well.
